quickwit/quickwit-actors/src/supervisor.rs

```diff
@@ -0,0 +1,383 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::time::Duration;
+
+use async_trait::async_trait;
+use tracing::{info, warn};
+
+use crate::mailbox::Inbox;
+use crate::{
+    Actor, ActorContext, ActorExitStatus, ActorHandle, ActorState, Handler, Health, KillSwitch,
+    Mailbox, Supervisable,
+};
+
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq)]
+pub struct SupervisorState {
+    num_panics: usize,
+    num_errors: usize,
+    num_kills: usize,
+}
+
+pub struct Supervisor<A: Actor> {
+    pub(crate) actor_name: String,
+    pub(crate) actor_factory: Box<dyn Fn() -> A + Sync + Send>,
+    pub(crate) inbox: Inbox<A>,
+    pub(crate) mailbox: Mailbox<A>,
+    pub(crate) handle_opt: Option<ActorHandle<A>>,
+    pub(crate) state: SupervisorState,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct SuperviseLoop;
+
+#[async_trait]
+impl<A: Actor> Actor for Supervisor<A> {
+    type ObservableState = SupervisorState;
+
+    fn observable_state(&self) -> Self::ObservableState {
+        self.state
+    }
+
+    fn name(&self) -> String {
+        format!("Supervisor({})", self.actor_name)
+    }
+
+    fn queue_capacity(&self) -> crate::QueueCapacity {
+        crate::QueueCapacity::Unbounded
+    }
+
+    async fn initialize(&mut self, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        ctx.schedule_self_msg(crate::HEARTBEAT, SuperviseLoop).await;
+        Ok(())
+    }
+
+    async fn finalize(
+        &mut self,
+        exit_status: &ActorExitStatus,
+        _ctx: &ActorContext<Self>,
+    ) -> anyhow::Result<()> {
+        match exit_status {
+            ActorExitStatus::Quit => {
+                if let Some(handle) = self.handle_opt.take() {
+                    handle.quit().await;
+                }
+            }
+            ActorExitStatus::Killed => {
+                if let Some(handle) = self.handle_opt.take() {
+                    handle.activate_kill_switch();
+                }
+            }
+            ActorExitStatus::Failure(_)
+            | ActorExitStatus::Success
+            | ActorExitStatus::DownstreamClosed => {}
+            ActorExitStatus::Panicked => {}
+        }
+
+        Ok(())
+    }
+}
+
+impl<A: Actor> Supervisor<A> {
+    async fn supervise(
+        &mut self,
+        ctx: &ActorContext<Supervisor<A>>,
+    ) -> Result<(), ActorExitStatus> {
+        match self.handle_opt.as_ref().unwrap().health() {
```

Comment:
It's possible to create a supervisor without a handle so unwrap.
We could either force the instantiation of a supervisor with an ActorHandle directly (and not Option<ActorHandle>) or avoid the unwrap here.

Reply:
The Option  is not for the instantiation, It's because joininig the Handle consume the Handle

Reply:
(in finalize)

Reply:
yes yes we need to keep the option ofc. I was just unsatisfied by the unwrap.

Reply:
(and we consume also the handle in the supervise method)

Reply:
To be complete, this is what I meant by enforcing the user to give a handle

```rust
impl<A: Actor> Supervisor<A> {
  pub fn new(
    actor_name: String,
    actor_factory: Box<dyn Fn() -> A + Sync + Send>,
    inbox: Inbox<A>,
    mailbox: Mailbox<A>,
    handle: ActorHandle<A>,
  ) -> {
    Self {
        actor_name,
        actor_factory,
        inbox,
        mailbox,
        handle_opt: Some(handle),
    }
 }
}
```

Reply:
ok I added a constructor.